src/stream/src/executor/sync_kv_log_store.rs

```diff
@@ -0,0 +1,1045 @@
+            state_store.update_vnode_bitmap(vnode_bitmap.clone());
+            serde.update_vnode_bitmap(vnode_bitmap.clone());
+            buffer
+                .buffer
+                .push_back((epoch, LogStoreBufferItem::UpdateVnodes(vnode_bitmap)));
+        }
+
+        *seq_id = FIRST_SEQ_ID;
+        Ok(())
+    }
+
+    async fn write_chunk(
+        metrics: &KvLogStoreMetrics,
+        serde: &LogStoreRowSerde,
+        start_seq_id: SeqIdType,
+        end_seq_id: SeqIdType,
+        buffer: &mut SyncedLogStoreBuffer,
+        chunk: StreamChunk,
+        state_store: &mut S::Local,
+    ) -> StreamExecutorResult<()> {
+        let chunk_to_flush =
+            { buffer.add_or_flush_chunk(start_seq_id, end_seq_id, chunk, state_store) };
+        match chunk_to_flush {
+            None => {}
+            Some(chunk_to_flush) => {
+                let new_vnode_bitmap = flush_chunk(
+                    metrics,
+                    serde,
+                    start_seq_id,
+                    end_seq_id,
+                    state_store,
+                    chunk_to_flush,
+                )
+                .await?;
+                {
+                    buffer.add_flushed_item_to_buffer(
+                        start_seq_id,
+                        end_seq_id,
+                        new_vnode_bitmap,
+                        state_store.epoch(),
+                    );
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+struct SyncedLogStoreBuffer {
+    buffer: VecDeque<(u64, LogStoreBufferItem)>,
+    max_size: usize,
+    next_chunk_id: ChunkId,
+    metrics: KvLogStoreMetrics,
+}
+
+async fn flush_chunk(
+    metrics: &KvLogStoreMetrics,
+    serde: &LogStoreRowSerde,
+    start_seq_id: SeqIdType,
+    end_seq_id: SeqIdType,
+    state_store: &mut impl LocalStateStore,
+    chunk: StreamChunk,
+) -> StreamExecutorResult<Bitmap> {
+    tracing::trace!("Flushing chunk: start_seq_id: {start_seq_id}, end_seq_id: {end_seq_id}");
+    let epoch = state_store.epoch();
+    let mut vnode_bitmap_builder = BitmapBuilder::zeroed(serde.vnodes().len());
+    let mut flush_info = FlushInfo::new();
+    for (i, (op, row)) in chunk.rows().enumerate() {
+        let seq_id = start_seq_id + (i as SeqIdType);
+        assert!(seq_id <= end_seq_id);
+        let (vnode, key, value) = serde.serialize_data_row(epoch, seq_id, op, row);
+        vnode_bitmap_builder.set(vnode.to_index(), true);
+        flush_info.flush_one(key.estimated_size() + value.estimated_size());
+        state_store.insert(key, value, None)?;
+    }
+    flush_info.report(metrics);
+    state_store.flush().await?;
+
+    Ok(vnode_bitmap_builder.finish())
+}
+
+impl SyncedLogStoreBuffer {
+    fn add_or_flush_chunk(
+        &mut self,
+        start_seq_id: SeqIdType,
+        end_seq_id: SeqIdType,
+        chunk: StreamChunk,
+        state_store: &mut impl LocalStateStore,
```

Comment:
Seems that the `state_store` is passed only to call `state_store.epoch()`. We'd better pass the barrier epoch directly.